File: leetcode/strings/AddBinary.java
public class AddBinary {
    
    public static String addBinary(String a, String b) {
        StringBuilder s = new StringBuilder();

        int idxa = a.length()-1;
        int idxb = b.length()-1;
        int carry = 0;
        
        while(idxa>=0 || idxb>=0) {
            int bita = idxa >= 0 ? ((int) a.charAt(idxa--)) - '0' : 0;
            int bitb = idxb >= 0 ? ((int) b.charAt(idxb--)) - '0' : 0;

            int sum = bita ^ bitb ^ carry;
            carry = (bita + bitb + carry)>1 ? 1 : 0;
            s.insert(0, sum);
        }
        if (carry==1) s.insert(0, carry);

        return s.toString();
    }

    public static void main(String[] args) {
        String s = addBinary("1010", "1011");
        System.out.println(s);
    }
}

File: leetcode/strings/two_pointers/add_binary.py
"""
67. Add Binary
Solved
Easy
Topics
Companies
Given two binary strings a and b, return their sum as a binary string.
"""


class AddBinary(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        i, j = len(a) - 1, len(b) - 1
        sum, carry = 0, 0
        res = ""

        while i >= 0 or j >= 0:
            bitA = 0 if i < 0 else ord(a[i]) - ord("0")
            bitB = 0 if j < 0 else ord(b[j]) - ord("0")

            total = (bitA + bitB + carry) % 2
            carry = (bitA + bitB + carry) // 2
            res = str(total) + res

            i, j = i - 1, j - 1

        if carry:
            res = "1" + res

        return res
